src/main/flight/gps_rescue.c

```diff
@@ -0,0 +1,373 @@
+        previousTimeUs = currentTimeUs;
+    }
+
+    rescueState.sensor.accMagnitude = (float) sqrt(sq(ABS((acc.accADC[Z] / acc.dev.acc_1G))) + sq(ABS((acc.accADC[X] / acc.dev.acc_1G))) + sq(ABS((acc.accADC[Y] / acc.dev.acc_1G))));
+    rescueState.sensor.accMagnitudeAvg = (rescueState.sensor.accMagnitudeAvg * 0.998f) + (rescueState.sensor.accMagnitude * 0.002f);
+}
+
+void performSanityChecks()
+{
+    if (rescueState.phase == RESCUE_IDLE) {
+        rescueState.failure = RESCUE_HEALTHY;
+        
+        return;
+    }
+
+    // Do not abort until each of these items is fully tested
+    if (rescueState.failure != RESCUE_HEALTHY) {
+        if (gpsRescue()->sanityChecks == RESCUE_SANITY_ON 
+            || (gpsRescue()->sanityChecks == RESCUE_SANITY_FS_ONLY && rescueState.isFailsafe == true)) {
+            rescueState.phase = RESCUE_ABORT;
+        }
+    }
+
+    // Check if crash recovery mode is active, disarm if so.
+    if (crashRecoveryModeActive()) {
+        rescueState.failure = RESCUE_CRASH_DETECTED;
+    }
+
+    //  Things that should run at a low refresh rate (such as flyaway detection, etc)
+    //  This runs at ~1hz
+
+    static uint32_t previousTimeUs;
+
+    const uint32_t currentTimeUs = micros();
+    const uint32_t dTime = currentTimeUs - previousTimeUs;
+
+    if (dTime < 1E6) {
+        return;
+    }
+
+    previousTimeUs = currentTimeUs;
+
+     // Stalled movement detection
+    static int8_t gsI = 0;
+
+    gsI = constrain(gsI + (rescueState.sensor.groundSpeed < 150) ? 1 : -1, -10, 10);
+
+    if (gsI == 10) {
+        rescueState.failure = RESCUE_CRASH_DETECTED;
+    }
+
+    // Minimum sat detection
+    static int8_t msI = 0;
+
+    msI = constrain(msI + (rescueState.sensor.numSat < gpsRescue()->minSats) ? 1 : -1, -5, 5);
+
+    if (msI == 5) {
+        rescueState.failure = RESCUE_FLYAWAY;
+    }
+
+    // Minimum distance detection (100m)
+    if (rescueState.sensor.distanceToHome < 100) {
+        rescueState.failure = RESCUE_TOO_CLOSE;
+    }
+}
+
+void rescueStart()
+{
+    rescueState.phase = RESCUE_INITIALIZE;
+}
+
+void rescueStop()
+{
+    rescueState.phase = RESCUE_IDLE;
+}
+
+// Things that need to run regardless of GPS rescue mode being enabled or not
+void idleTasks()
+{
+    gpsRescueAngle[AI_PITCH] = 0;
+    gpsRescueAngle[AI_ROLL] = 0;
+
+    // Store the max altitude we see not during RTH so we know our fly-back minimum alt
+    rescueState.sensor.maxAltitude = MAX(rescueState.sensor.currentAltitude, rescueState.sensor.maxAltitude);
+    // Store the max distance to home during normal flight so we know if a flyaway is happening
+    rescueState.sensor.maxDistanceToHome = MAX(rescueState.sensor.distanceToHome, rescueState.sensor.maxDistanceToHome);
+
+    rescueThrottle = rcCommand[THROTTLE];
+
+    //to do: have a default value for hoverThrottle
+    float ct = getCosTiltAngle();
+    if (ct > 0.5 && ct < 0.96 && throttleSamples < 1E6 && rescueThrottle > 1070) { //5 to 45 degrees tilt
```

Comment:
Since throttle is non linear, we wanted to sample the values when the craft is greater than the minimum angle we use during our RTH.  Rescue mode never goes below 5 degrees, because at that point you start poisoning the GPS derived compass heading.